include/nbl/builtin/glsl/ext/FFT/fft.glsl

```diff
@@ -0,0 +1,328 @@
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_MAX_ITEMS_PER_THREAD
+#error "_NBL_GLSL_EXT_FFT_MAX_ITEMS_PER_THREAD should be defined."
+#endif
+
+
+// TODO: Investigate why +1 solves all glitches
+#define _NBL_GLSL_EXT_FFT_SHARED_SIZE_NEEDED_ _NBL_GLSL_EXT_FFT_MAX_DIM_SIZE_ + 1
+
+#ifdef _NBL_GLSL_SCRATCH_SHARED_DEFINED_
+    #if NBL_GLSL_LESS(_NBL_GLSL_SCRATCH_SHARED_SIZE_DEFINED_,_NBL_GLSL_EXT_FFT_SHARED_SIZE_NEEDED_)
+        #error "Not enough shared memory declared for ext::FFT !"
+    #endif
+#else
+    #if NBL_GLSL_GREATER(_NBL_GLSL_EXT_FFT_SHARED_SIZE_NEEDED_,0)
+        #define _NBL_GLSL_SCRATCH_SHARED_DEFINED_ nbl_glsl_fftScratchShared
+        #define _NBL_GLSL_SCRATCH_SHARED_SIZE_DEFINED_ _NBL_GLSL_EXT_FFT_SHARED_SIZE_NEEDED_
+        shared uint _NBL_GLSL_SCRATCH_SHARED_DEFINED_[_NBL_GLSL_EXT_FFT_SHARED_SIZE_NEEDED_];
+    #endif
+#endif
+
+// Push Constants
+
+#define _NBL_GLSL_EXT_FFT_DIRECTION_X_ 0
+#define _NBL_GLSL_EXT_FFT_DIRECTION_Y_ 1
+#define _NBL_GLSL_EXT_FFT_DIRECTION_Z_ 2
+
+#define _NBL_GLSL_EXT_FFT_CLAMP_TO_EDGE_ 0
+#define _NBL_GLSL_EXT_FFT_FILL_WITH_ZERO_ 1
+
+#ifndef _NBL_GLSL_EXT_FFT_PUSH_CONSTANTS_DEFINED_
+#define _NBL_GLSL_EXT_FFT_PUSH_CONSTANTS_DEFINED_
+layout(push_constant) uniform PushConstants
+{
+    layout (offset = 0) uvec3 dimension;
+    layout (offset = 16) uvec3 padded_dimension;
+	layout (offset = 32) uint direction;
+    layout (offset = 36) uint is_inverse;
+    layout (offset = 40) uint padding_type; // clamp_to_edge or fill_with_zero
+} pc;
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_GET_DATA_DECLARED_
+#define _NBL_GLSL_EXT_FFT_GET_DATA_DECLARED_
+vec2 nbl_glsl_ext_FFT_getData(in uvec3 coordinate, in uint channel);
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_SET_DATA_DECLARED_
+#define _NBL_GLSL_EXT_FFT_SET_DATA_DECLARED_
+void nbl_glsl_ext_FFT_setData(in uvec3 coordinate, in uint channel, in vec2 complex_value);
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_GET_DATA_DEFINED_
+#error "You need to define `nbl_glsl_ext_FFT_getData` and mark `_NBL_GLSL_EXT_FFT_GET_DATA_DEFINED_`!"
+#endif
+#ifndef _NBL_GLSL_EXT_FFT_SET_DATA_DEFINED_
+#error "You need to define `nbl_glsl_ext_FFT_setData` and mark `_NBL_GLSL_EXT_FFT_SET_DATA_DEFINED_`!"
+#endif
+
+// Count Leading Zeroes (naive?)
+uint clz(in uint x) 
+{
+    uint n = 0;
+	if (x == 0) { return 32; }
+    if (x <= 0x0000ffff) { n += 16; x <<= 16; }
+    if (x <= 0x00ffffff) { n +=  8; x <<= 8; }
+    if (x <= 0x0fffffff) { n +=  4; x <<= 4; }
+    if (x <= 0x3fffffff) { n +=  2; x <<= 2; }
+    if (x <= 0x7fffffff) { n++; };
+    return n;
+}
+
+uint reverseBits(in uint x)
+{
+	uint count = 4 * 8 - 1;
+	uint reverse_num = x; 
+
+    x >>= 1;  
+    while(x > 0) 
+    { 
+       reverse_num <<= 1;        
+       reverse_num |= x & 1; 
+       x >>= 1; 
+       count--; 
+    } 
+    reverse_num <<= count;
+    return reverse_num; 
+}
+
+uint calculate_twiddle_power(in uint threadId, in uint iteration, in uint logTwoN, in uint N) 
```

Comment:
every single function you ever define in a GLSL builtin needs the correct prefix to the name (in this case `nbl_glsl_ext_FFT_`)... otherwise chaos ensues

Reply:
also wouldn't your life be easier if this function just took `log2N` without `N` at all?